File: MatrixShader/src/MatrixShader.Core/Services/MachineFingerprint.cs
using System.Security.Cryptography;
using System.Text;

namespace MatrixShader.Core.Services;

/// <summary>
/// Generates a deterministic machine fingerprint for activation tracking.
/// SHA256 of (MachineName + UserName + OSVersion), truncated to 16 hex chars.
/// </summary>
public static class MachineFingerprint
{
    private static string? _cached;

    public static string Get()
    {
        if (_cached != null)
            return _cached;

        var raw = string.Join("|",
            Environment.MachineName,
            Environment.UserName,
            Environment.OSVersion.ToString());

        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(raw));
        _cached = Convert.ToHexString(hash)[..16].ToLowerInvariant();
        return _cached;
    }
}

File: linux/machine_fingerprint.py
"""Deterministic machine fingerprint for license activation tracking.

Port of MatrixShader.Core/Services/MachineFingerprint.cs.
SHA256 of (hostname | username | OS version), truncated to 16 hex chars.
"""

import hashlib
import os
import platform

_cached = None


def get():
    """Return a 16-char hex fingerprint for this machine."""
    global _cached
    if _cached is not None:
        return _cached

    raw = "|".join([
        platform.node(),           # hostname (Environment.MachineName)
        os.getenv("USER", os.getenv("USERNAME", "")),  # Environment.UserName
        platform.platform(),       # Environment.OSVersion.ToString()
    ])

    digest = hashlib.sha256(raw.encode("utf-8")).hexdigest()
    _cached = digest[:16]
    return _cached
